File: auto-instrumentation/okhttp/okhttp-3.0/library/src/main/java/io/opentelemetry/instrumentation/library/okhttp/v3_0/internal/TracingCallback.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.library.okhttp.v3_0.internal;

import io.opentelemetry.context.Context;
import io.opentelemetry.context.Scope;
import java.io.IOException;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Response;

/**
 * This class is internal and is hence not for public use. Its APIs are unstable and can change at
 * any time.
 */
public final class TracingCallback implements Callback {
    private final Callback delegate;
    private final Context callingContext;

    public TracingCallback(Callback delegate, Context callingContext) {
        this.delegate = delegate;
        this.callingContext = callingContext;
    }

    @Override
    public void onFailure(Call call, IOException e) {
        try (Scope scope = callingContext.makeCurrent()) {
            delegate.onFailure(call, e);
        }
    }

    @Override
    public void onResponse(Call call, Response response) throws IOException {
        try (Scope scope = callingContext.makeCurrent()) {
            delegate.onResponse(call, response);
        }
    }
}

File: instrumentation/okhttp3/library/src/main/java/io/opentelemetry/instrumentation/library/okhttp/v3_0/internal/TracingCallback.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.library.okhttp.v3_0.internal

import io.opentelemetry.context.Context
import java.io.IOException
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Response

/**
 * This class is internal and is hence not for public use. Its APIs are unstable and can change at
 * any time.
 */
internal class TracingCallback(
    private val delegate: Callback,
    private val callingContext: Context
) : Callback {

    override fun onFailure(call: Call, e: IOException) {
        callingContext.makeCurrent().use {
            delegate.onFailure(call, e)
        }
    }

    @Throws(IOException::class)
    override fun onResponse(call: Call, response: Response) {
        callingContext.makeCurrent().use {
            delegate.onResponse(call, response)
        }
    }
}
